core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -169,35 +174,108 @@ protected PlanNodeStatsEstimate visitLogicalExpression(LogicalExpression node, V
 
         private PlanNodeStatsEstimate estimateLogicalAnd(List<Expression> terms)
         {
-            // first try to estimate in the fair way
-            PlanNodeStatsEstimate estimate = process(terms.get(0));
-            if (!estimate.isOutputRowCountUnknown()) {
-                for (int i = 1; i < terms.size(); i++) {
-                    estimate = new FilterExpressionStatsCalculatingVisitor(estimate, session, types).process(terms.get(i));
-
-                    if (estimate.isOutputRowCountUnknown()) {
-                        break;
-                    }
-                }
+            List<PlanNodeStatsEstimate> estimates = estimateCorrelatedExpressions(terms);
+            // Sort known estimates by selectivity
+            List<PlanNodeStatsEstimate> knownSortedEstimates = estimates.stream()
+                    .filter(estimate -> !estimate.isOutputRowCountUnknown())
+                    .sorted(Comparator.comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
+                    .collect(toImmutableList());
 
-                if (!estimate.isOutputRowCountUnknown()) {
-                    return estimate;
+            if (knownSortedEstimates.isEmpty()) {
+                return PlanNodeStatsEstimate.unknown();
+            }
+
```

Comment:
you probably should return early when `input.getOutputRowCount() == 0 or PlanNodeStatsEstimate#isOutputRowCountUnknown` (unless it's already checked before calling `estimateLogicalAnd`). Please add a test for that

Reply:
Unknown case was already fin eight existing code, I've updated `FilterExpressionStatsCalculatingVisitor#process` to handle the zero input case. Also added tests.